faunadb-scala/src/main/scala/faunadb/values/CodecMacro.scala

```diff
@@ -0,0 +1,64 @@
+package faunadb.values
+
+import scala.reflect.macros.blackbox
+
+class CodecMacro(val c: blackbox.Context) {
+  import c.universe._
+
+  def caseClassImpl[T: c.WeakTypeTag]: c.Tree = {
+    val tag = weakTypeOf[T]
+
+    q""" new Codec[$tag] {
+      override def decode(value: Value, path: FieldPath): Result[$tag] =
+        ${getDecodedObject(tag)}
+
+      override def encode(value: $tag): Value =
+        ${getEncodedObject(tag)}
+    } """
+  }
+
+  def getEncodedObject(tag: c.Type): c.Tree = {
+    val fields = getFields(tag).map { field =>
+      if (field._2.typeConstructor =:= weakTypeOf[Option[_]].typeConstructor)
+        q""" ${varName(field).toString} -> (value.${varName(field).toTermName}.map(v => v: Value).getOrElse(NullV)) """
+      else
+        q""" ${varName(field).toString} -> (value.${varName(field).toTermName}: Value) """
+    }
+
+    q"ObjectV.apply(..$fields)"
+  }
+
+  def getDecodedObject(tag: c.Type): c.Tree = {
+    val fields = getFields(tag)
+
+    val fieldsFragments = fields.map { field =>
+      if (field._2.typeConstructor =:= weakTypeOf[Option[_]].typeConstructor)
+        fq"${varName(field)} <- VSuccess(value(${varName(field).toString}).to[${field._2.typeArgs.head}].toOpt, ${varName(field).toString})"
+      else
+        fq"${varName(field)} <- value(${varName(field).toString}).to[${field._2}]"
+    }
+
+    q"for (..$fieldsFragments) yield new $tag(..${fields.map(varName)})"
+  }
+
+  def varName(field: (c.Symbol, c.Type)): c.Name =
+    field._1.name.decodedName
+
+  def getFields(tag: c.Type): List[(c.Symbol, c.Type)] = {
+    if (tag.typeSymbol.isClass) {
```

Comment:
I'd say combine these conditionals w/ boolean logic... our scala style is to always require braces for multi-line if branches.